### PR TITLE
[MXNET -1030] Cosine Embedding Loss

### PR DESCRIPTION
**Pull request by gaurav-gireesh**
## Description ##
For a target label 1 or -1, vectors target and prediction, the function computes the cosine distance between the vectors. This can be interpreted as how similar/dissimilar two input vectors are.
Can be used as a loss function while training models.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)

- [x] Code is well-documented: 

- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Cosine Embedding Loss computation function tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.

### CODE REVIEW

**Comment by vrakesh** [comment]
@mxnet-label-bot [pr-awaiting-review]

File: python/mxnet/gluon/loss.py [hunk 1]
```diff
@@ -706,3 +706,53 @@ def hybrid_forward(self, F, pred, positive, negative):
                      axis=self._batch_axis, exclude=True)
         loss = F.relu(loss + self._margin)
         return _apply_weighting(F, loss, self._weight, None)
+
+class CosineEmbeddingLoss(Loss):
+    r"""For a target label 1 or -1, vectors target and pred, the function computes the cosine distance
+    between the vectors. This can be interpretted as how similar/dissimilar two input vectors are.
+
+
+    `pred`, `target` can have arbitrary shape as long as they have the same number of elements.
```
Comment (szha) [inline python/mxnet/gluon/loss.py]: could you document the formula here? our website supports mathjax so you can write latex for it.

Comment (gaurav-gireesh) [reply python/mxnet/gluon/loss.py]: @szha Thank you! Incorporating that in the documentation.

Comment (szha) [reply python/mxnet/gluon/loss.py]: The rendering needs to be fixed:
http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-12750/2/api/python/gluon/loss.html#mxnet.gluon.loss.CosineEmbeddingLoss
Also, could you list the loss in this file? https://github.com/apache/incubator-mxnet/blame/master/docs/api/python/gluon/loss.md#L14

File: python/mxnet/gluon/loss.py [hunk 2]
```diff
@@ -706,3 +706,77 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+        pred = _reshape_like(F, pred, target)
+        cos_sim = self.cosine_similarity(F, pred, target)
+        y_1 = label == 1
+        y_minus_1 = label == -1
+        cos_sim_a = (1 - cos_sim) * y_1
+        cos_sim_b = F.broadcast_maximum(F.array([0]), y_minus_1 * (cos_sim - self._margin), axis=1)
+        return cos_sim_a + cos_sim_b
+
+    def cosine_similarity(self, F, F1, F2, axis=-1):
+        F1_norm = F1.norm(axis=axis).reshape(-1, 1)
```
Comment (roywei) [inline python/mxnet/gluon/loss.py]: use F.norm(x, axis=axis)... to support hybridization

File: python/mxnet/gluon/loss.py [hunk 3]
```diff
@@ -706,3 +706,77 @@ def hybrid_forward(self, F, pred, positive, negative):
                      axis=self._batch_axis, exclude=True)
         loss = F.relu(loss + self._margin)
         return _apply_weighting(F, loss, self._weight, None)
+
+class CosineEmbeddingLoss(Loss):
+    r"""For a target label 1 or -1, vectors target and pred, the function computes the cosine distance
+    between the vectors. This can be interpretted as how similar/dissimilar two input vectors are.
+
+    .. math::
+        L = \begin{gather*}
```
Comment (anirudhacharya) [inline python/mxnet/gluon/loss.py]: isn't the loss a cumulative sum over all data points? so the formula should begin with `\sum_i`

File: tests/python/unittest/test_loss.py [hunk 4]
```diff
@@ -348,6 +348,22 @@ def test_triplet_loss():
...
+    target = mx.nd.array([[1, 1, 1, 1],
+                          [1, 2, 3, 4]])
+    Loss = gluon.loss.CosineEmbeddingLoss()
+    loss = Loss(pred, target, label)
+
+    #computing numpy way
+    numerator = mx.nd.sum(pred * target, keepdims=True, axis=1)
+    denominator = mx.nd.sqrt(mx.nd.sum(pred**2, axis=1, keepdims=True)) \
+    * mx.nd.sqrt(mx.nd.sum(target**2, axis=1, keepdims=True))
+    assert_almost_equal(loss.asnumpy(), (1-numerator/denominator).asnumpy())
```
Comment (anirudhacharya) [inline tests/python/unittest/test_loss.py]: why not test using the Module API with `mod.score` and `mod.fit`

File: python/mxnet/gluon/loss.py (hunk 1 above)
Comment (szha) [reply python/mxnet/gluon/loss.py]: Latest doc: http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-12750/6/api/python/gluon/loss.html#mxnet.gluon.loss.CosineEmbeddingLoss
It requires an empty line after `.. math::` to render properly

Comment (roywei) [reply python/mxnet/gluon/loss.py]: Please refer to[ huber loss ](https://github.com/gaurav-gireesh/incubator-mxnet/blob/cosineloss/python/mxnet/gluon/loss.py#L478)and use only `{case}`, remove `{gather}`. maybe your mathjax version and what we used in the website is different, so some formula is not recognized. A safe way is to use only what have been correctly rendered, those should be more than enough

File: tests/python/unittest/test_loss.py [hunk 5]
```diff
@@ -348,6 +348,22 @@ def test_triplet_loss():
             optimizer='adam')
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.05
 
+@with_seed()
```
Comment (sandeep-krishnamurthy) [inline tests/python/unittest/test_loss.py]: Please add additional tests for label = -1 and hybridization.

File: python/mxnet/gluon/loss.py [hunk 6]
```diff
@@ -706,3 +706,57 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+    def hybrid_forward(self, F, pred, target, label):
+        pred = _reshape_like(F, pred, target)
+        cos_sim = self.cosine_similarity(F, pred, target)
+        y_1 = label == 1
+        y_minus_1 = label == -1
+        cos_sim_a = (1 - cos_sim) * y_1
+        cos_sim_b = F.broadcast_maximum(F.array([0]), y_minus_1 * (cos_sim - self._margin), axis=1)
+        return cos_sim_a + cos_sim_b
+
+    def cosine_similarity(self, F, x, y, axis=-1):
```
Comment (sandeep-krishnamurthy) [inline python/mxnet/gluon/loss.py]: nit: make it internal by prepending with _ for function name.

**Comment by sandeep-krishnamurthy** [comment]
Nice first contribution. Welcome to MXNet community @gaurav-gireesh

**Comment by gaurav-gireesh** [comment]
> Nice first contribution. Welcome to MXNet community @gaurav-gireesh

Thank you Sandeep! Looking forward to collaborate and learn together with the community!

File: tests/python/unittest/test_loss.py [hunk 7]
```diff
@@ -348,7 +348,35 @@ def test_triplet_loss():
             optimizer='adam')
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.05
 
-
+@with_seed()
+def test_cosine_loss():
+    #For similarity check
+    label = mx.nd.array([1])
+    pred = mx.nd.array([[1, 1, 1, 1],
```
Comment (roywei) [inline tests/python/unittest/test_loss.py]: Please also rename params in test to input1, input2, label to avoid confusion

File: tests/python/unittest/test_loss.py [hunk 8]
```diff
@@ -348,7 +348,35 @@ def test_triplet_loss():
...
+    Loss = gluon.loss.CosineEmbeddingLoss()
+    loss = Loss(pred, target, label)
+
+    #computing numpy way
+    numerator = mx.nd.sum(pred * target, keepdims=True, axis=1)
+    denominator = mx.nd.sqrt(mx.nd.sum(pred**2, axis=1, keepdims=True)) \
+    * mx.nd.sqrt(mx.nd.sum(target**2, axis=1, keepdims=True))
+    assert_almost_equal(loss.asnumpy(), (1-numerator/denominator).asnumpy())
+
+    # Label == -1
```
Comment (roywei) [inline tests/python/unittest/test_loss.py]: Please refer to pytorch unit test: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L4986
Create random input1 input2 and labels, labels can be randomly 1 or -1, so you can test them at once. No need to repeat with just label = -1 on fixed inputs.

File: python/mxnet/gluon/loss.py [hunk 9]
```diff
@@ -706,3 +706,58 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+    weight : float or None
+        Global scalar weight for loss.
+    batch_axis : int, default 0
+        The axis that represents mini-batch.
+    margin : float
+        Margin of separation between correct and incorrect pair.
+
+
+    Inputs:
+        - **input1**: a tensor with arbitrary shape
```
Comment (roywei) [inline python/mxnet/gluon/loss.py]: Update document on param names

File: python/mxnet/gluon/loss.py [hunk 10]
```diff
@@ -706,3 +706,58 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+    r"""For a target label 1 or -1, vectors target and pred, the function computes the cosine distance
+    between the vectors. This can be interpretted as how similar/dissimilar two input vectors are.
+
+    .. math::
+
+        L = \sum_i \begin{cases} 1 - {cos\_sim({input1}_i, {input2}_i)} & \text{ if } {label}_i = 1\\
+                         {cos\_sim({input1}_i, {input2}_i)} & \text{ if } {label}_i = -1 \end{cases}\\
+        cos\_sim(input1, input2) = \frac{{input1}_i.{input2}_i}{||{input1}_i||.||{input2}_i||}
+
+    `pred`, `target` can have arbitrary shape as long as they have the same number of elements.
```
Comment (roywei) [inline python/mxnet/gluon/loss.py]: Update document on param names

File: tests/python/unittest/test_loss.py (hunk 5 above)
Comment (gaurav-gireesh) [reply tests/python/unittest/test_loss.py]: > Please add additional tests for label = -1 and hybridization.

1. Added tests for label 1 and -1 in a randomly generated set of labels,
2. This function is meant to be utility similar to cosine_distance.

File: python/mxnet/gluon/loss.py [hunk 11]
```diff
@@ -767,3 +767,69 @@ def hybrid_forward(self, F, pred, target, sample_weight=None, epsilon=1e-08):
...
+        The axis that represents mini-batch.
+    margin : float
+        Margin of separation between correct and incorrect pair.
+
+
+    Inputs:
+        - **input1**: a tensor with arbitrary shape
+        - **input2**: another tensor with same shape as pred to which input1 is
+          compared for similarity and loss calculation
+        - **sample_weight**: element-wise weighting tensor. Must be broadcastable
```
Comment (szha) [inline python/mxnet/gluon/loss.py]: This needs to be added to hybrid_forward. Could you also put this after `label`?

File: python/mxnet/gluon/loss.py [hunk 12]
```diff
@@ -767,3 +767,71 @@ def hybrid_forward(self, F, pred, target, sample_weight=None, epsilon=1e-08):
...
+          to the same shape as input1. For example, if input1 has shape (64, 10)
+          and you want to weigh each sample in the batch separately,
+          sample_weight should have shape (64, 1).
+        - **label**: A 1-D tensor indicating for each pair input1 and input2, target label is 1 or -1
```
Comment (szha) [inline python/mxnet/gluon/loss.py]: Label is not the last input variable. It's before the sample_weight, and the second last variable

File: python/mxnet/gluon/loss.py [hunk 13]
```diff
@@ -767,3 +767,71 @@ def hybrid_forward(self, F, pred, target, sample_weight=None, epsilon=1e-08):
             loss += stirling_factor
         loss = _apply_weighting(F, loss, self._weight, sample_weight)
         return F.mean(loss)
+
+
+class CosineEmbeddingLoss(Loss):
+    r"""For a target label 1 or -1, vectors target and pred, the function computes the cosine distance
+    between the vectors. This can be interpretted as how similar/dissimilar two input vectors are.
```
Comment (sandeep-krishnamurthy) [inline python/mxnet/gluon/loss.py]: nit: interpreted

File: python/mxnet/gluon/loss.py (hunk 11 above)
Comment (gaurav-gireesh) [reply python/mxnet/gluon/loss.py]: Addressed this. Thanks

**Comment by roywei** [comment]
@mxnet-label-bot [pr-awaiting-merge]

File: tests/python/unittest/test_loss.py [hunk 14]
```diff
@@ -404,6 +421,7 @@ def test_poisson_nllloss_mod():
             optimizer='adam')
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.05
 
+
```
[the remark this answers is not in the document]
Comment (gaurav-gireesh) [reply tests/python/unittest/test_loss.py]: Yes. It is not required. Following the other test modules. They have a 2 lines gap after the last test function and main function.

**Comment by anirudhacharya** [comment]
LGTM

**Comment by sandeep-krishnamurthy** [comment]
@szha - This is good to go after your review.

**Comment by gaurav-gireesh** [comment]
@mxnet-label-bot remove [pr-awaiting-review]
